jobs/AnalyzeResults.php

```diff
@@ -30,90 +34,191 @@ public function __destruct() {
+        $f1Score = $this->calculateF1Score( $precision, $recall );
+
+        $precisionAt10 = $this->calculatePrecision( $truePositivesAt10, $falsePositivesAt10 );
+        $precisionAt25 = $this->calculatePrecision( $truePositivesAt25, $falsePositivesAt25 );
+        $precisionAt50 = $this->calculatePrecision( $truePositivesAt50, $falsePositivesAt50 );
+        $precisionAt100 = $this->calculatePrecision( $truePositivesAt100, $falsePositivesAt100 );
+
         fwrite(
             $this->out,
-            "ARITHMETIC MEAN," .
-            array_sum( $f1Scores ) / count( $resultsets) . "," .
-            array_sum( $precisionTop30s ) / count( $resultsets) .
-            "\n"
+            "OVERALL," .
+            // overall f1score is not an arithmetic average of the individual f1scores,
+            // but a calculation based on the harmonic means of all precision &
+            // recall values, so that it's more sensitive to extremes
+            $f1Score . "," .
+            ( $precisionAt10 ?? '/' ) . "," .
+            ( $precisionAt25 ?? '/' ) . "," .
+            ( $precisionAt50 ?? '/' ) . "," .
+            ( $precisionAt100 ?? '/' ) . "," .
+            ( $recall ?? '/' ) . "\n"
         );
     }
 
-    private function calculateF1Score( int $resultsetId, string $searchTerm ) : float {
-        $knownGoodImageCount = $this->db->query(
-            'select count(*) as count from results_by_component where ' .
-            'term = "' . $this->db->real_escape_string( trim( $searchTerm ) ) . '" and ' .
-            'rating=1'
-        )->fetch_object()->count;
-        $foundGoodImageCount = $this->db->query(
+    private function getTruePositive( int $resultsetId, int $offset ) : int {
+        return (int) $this->db->query(
             'select count(*) as count from labeledResult where ' .
-            'resultsetId = "' . intval( $resultsetId ) . '" and ' .
-            'rating=1'
+            'resultsetId = ' . intval( $resultsetId ) . ' and ' .
+            'position < ' . intval($offset) . ' and ' . // only count hits before a given offset
+            'score > 1 and ' . // only count results that were scored with this algorithm
+            'rating = 1'
         )->fetch_object()->count;
-        if ( $foundGoodImageCount == 0 ) {
-            return 0;
-        }
-        $foundBadImageCount = $this->db->query(
+    }
+
+    private function getFalsePositive( int $resultsetId, int $offset ) : int {
+        return (int) $this->db->query(
             'select count(*) as count from labeledResult where ' .
-            'resultsetId = "' . intval( $resultsetId ) . '" and ' .
-            'rating<1'
+            'resultsetId = ' . intval( $resultsetId ) . ' and ' .
+            'position < ' . intval($offset) . ' and ' . // only count hits before a given offset
+            'score > 1 and ' . // only count results that were scored with this algorithm
+            'rating = -1'
```

Comment:
Should this be `rating<1` ?

Reply:
This was intentional - `0` is "meh", and in my voting, I've been treating that as "well, it's probably not what I'd be looking for for that specific term, but it makes some sense". I.e., not a bad result.
Since it's also not a good result, I'm just completely ignoring those.

That said, I can be persuaded to go for "perfect or nothing" and even consider "meh" results as false positives.
What do you think?